### PR TITLE
test: integration tests for keplr login endpoint

### PR DESCRIPTION
**Pull request by wgwz**
## Description

Closes: https://github.com/regen-network/regen-registry/issues/1086

- Adds e2e/integration tests for the login and logout endpoints for the login system
- Adds a middleware function called `ensureLoggedIn` that will be used to protect endpoints that require authentication
- Adds a "utils" module for the e2e/integration tests for deduping code across tests

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] targeted `dev` branch
- [ ] provided a link to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] submit a PR against `master` branch after this one (and other PRs depending on this one, e.g. on regen-network/regen-web, if applicable) get merged

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 1]
```diff
@@ -10,8 +18,65 @@ describe('web3auth login endpoint', () => {
...
+    );
+    const resp = await fetch(req, {
+      body: JSON.stringify({ signature: 'FOOBAR' }),
+    });
+    expect(resp.status).toBe(500);
+  });
+
+  // note: if we need a key pair that comes from a mnemonic:
+  // https://github.com/chainapsis/keplr-wallet/blob/c6ea69512ee7487fe3dcd9ce89f928a96dbf44a2/packages/crypto/src/mnemonic.spec.ts#L6-L11
+  it('authenticates a new user successfully and creates a session...', async () => {
```
Comment (wgwz) [inline server/__tests__/e2e/web3auth.login.test.ts:42]: hey @blushi just wanted to get your eyes on this. this is the test that i was working on where i want to simulate the action of a keplr wallet. i've got a passing test for this after following some of the code in keplr. this test specifically tests the path where we have a new user that has never signed in to the application. i still have a few more paths to test, like authenticating a returning user, and then there are probably some more negative test cases to write. for the returning user test case, i was thinking that i'll generate and key pair from a mnemonic following keplr code and then the address would be the same each time. but i'm quite happy with this! WDYT?

Comment (blushi) [reply server/__tests__/e2e/web3auth.login.test.ts:65]: looks good!

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 2]
```diff
@@ -10,8 +24,157 @@ describe('web3auth login endpoint', () => {
...
+
+    const { response: loginResp, authHeaders } = await performLogin(
+      privKey,
+      pubKey,
+      signer,
+      nonce,
+    );
+    loginResponseAssertions(loginResp, signer);
+
+    // check that an authenticated user use an authenticated graphql query
```
Comment (flagrede) [inline server/__tests__/e2e/web3auth.login.test.ts:62]:
```suggestion
    // check that an authenticated user can use an authenticated graphql query
```

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 3]
```diff
@@ -10,8 +24,157 @@ describe('web3auth login endpoint', () => {
...
+    const resp = await fetch('http://localhost:5000/graphql', {
+      method: 'POST',
+      headers: authHeaders,
+      body: JSON.stringify({
+        query:
+          'mutation {getCurrentAddrs(input: {}) {clientMutationId results { addr } }}',
+      }),
+    });
+    const data = await resp.json();
+    // expect that the response contains the users current addresses
```
Comment (flagrede) [inline server/__tests__/e2e/web3auth.login.test.ts:72]:
```suggestion
    // expect that the response contains the user's current addresses
```

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 4]
```diff
@@ -10,8 +24,157 @@ describe('web3auth login endpoint', () => {
...
+    ]);
+  });
+
+  it('returns a permissions error if the user session is manipulated...', async () => {
+    const privKey = new PrivKeySecp256k1(
+      Mnemonic.generateWalletFromMnemonic(TEST_ACCOUNT_MNEMONIC),
+    );
+    const pubKey = privKey.getPubKey();
+    const signer = new Bech32Address(pubKey.getAddress()).toBech32('regen');
+    expect(signer).toBe('regen1hscq3r6zz9ucut2d0jqqdc9lqwvu8h47x73lvm');
```
Comment (flagrede) [inline server/__tests__/e2e/web3auth.login.test.ts:120]: Should the address be extracted since it is reused several times?

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 5]
```diff
@@ -10,8 +24,157 @@ describe('web3auth login endpoint', () => {
...
+          'mutation {getCurrentAddrs(input: {}) {clientMutationId results { addr } }}',
+      }),
+    });
+    const data = await resp.json();
+    expect(data).toHaveProperty('data.getCurrentAddrs.results', [
+      { addr: signer },
+    ]);
+  });
+
+  it('returns a permissions error if the user session is manipulated...', async () => {
```
Comment (flagrede) [inline server/__tests__/e2e/web3auth.login.test.ts:114]: Nice test 👏

**Review by flagrede — APPROVED** [review]
Looks good, nice work!

File: server/__tests__/e2e/web3auth.login.test.ts (hunk 5 above)
Comment (wgwz) [reply server/__tests__/e2e/web3auth.login.test.ts:114]: thanks! i felt this was an important enough implementation detail that i wanted to write a test for it :-)

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 6]
```diff
@@ -10,8 +25,157 @@ describe('web3auth login endpoint', () => {
...
+      body: JSON.stringify({
+        query:
+          'mutation {getCurrentAddrs(input: {}) {clientMutationId results { addr } }}',
+      }),
+    });
+    const data = await resp.json();
+    const errMsgs = data.errors.map(x => x.message);
+    // since the session cookie was manipulated, we expect that the user session is invalidated.
+    // the user session is invalidated is because the session is signed.
+    // because the session is signed with a secret that only the backend knows,
```
Comment (blushi) [inline server/__tests__/e2e/web3auth.login.test.ts:176]:
```suggestion
    // the user session is invalidated because the session is signed
    // with a secret that only the backend knows,
```

File: server/__tests__/e2e/web3auth.logout.test.ts [hunk 7]
```diff
@@ -9,9 +11,55 @@ describe('web3auth logout endpoint', () => {
...
     const resp = await fetch(req);
-    expect(resp.status !== 403).toBe(true);
+    expect(resp.status).toBe(401);
+  });
+
+  it('closes the user session and a user can no longer make authd requests', async () => {
+    // set up a key pair and sign the required login transaction..
+    const privKey = PrivKeySecp256k1.generateRandomKey();
+    const pubKey = privKey.getPubKey();
+    const signer = new Bech32Address(pubKey.getAddress()).toBech32('regen');
```
Comment (blushi) [inline server/__tests__/e2e/web3auth.logout.test.ts:27]: maybe we could extract this logic into a reusable fn?

File: server/__tests__/utils.ts [hunk 8]
```diff
@@ -9,14 +22,139 @@ export async function fetchCsrf(): Promise<{ cookie: string; token: string }> {
...
 
-export async function CSRFRequest(endpoint: string): Promise<Request> {
+export async function CSRFRequest(
+  endpoint: RequestInfo,
+  method: string,
+): Promise<Request> {
   const { cookie, token } = await fetchCsrf();
   const request = new Request(endpoint, {
-    method: 'POST',
+    method: method,
```
Comment (blushi) [inline server/__tests__/utils.ts:31]:
```suggestion
    method,
```

File: server/middleware/keplrStrategy.ts [hunk 9]
```diff
@@ -20,16 +20,19 @@ export function KeplrStrategy(): CustomStrategy {
     let client: PoolClient;
     try {
       const { signature } = req.body;
+      console.log('signature', signature);
```
Comment (blushi) [inline server/middleware/keplrStrategy.ts:23]: should we drop the following `console.log`'s?
```suggestion
```

**Review by blushi — APPROVED** [review]
Awesome work, great to have those tests setup! Just a few nits, pre-approving.

File: server/__tests__/e2e/web3auth.logout.test.ts (hunk 7 above)
Comment (wgwz) [reply server/__tests__/e2e/web3auth.logout.test.ts:27]: i'm going to hold off on this for now, i might handle it in my other PR: https://github.com/regen-network/registry-server/pull/284
